src/sif/autocost.cc

```diff
@@ -408,32 +425,28 @@ Cost AutoCost::EdgeCost(const baldr::DirectedEdge* edge,
   // either the computed edge speed or optional top_speed
   auto edge_speed = tile->GetSpeed(edge, flow_mask_, seconds);
   auto final_speed = std::min(edge_speed, top_speed_);
-
-  float sec = (edge->length() * speedfactor_[final_speed]);
+  float sec = edge->length() * speedfactor_[final_speed];
 
   if (shortest_) {
     return Cost(edge->length(), sec);
   }
 
-  float factor =
-      (edge->use() == Use::kFerry)
-          ? ferry_factor_
-          : (edge->use() == Use::kRailFerry) ? rail_ferry_factor_ : density_factor_[edge->density()];
+  // base factor is either ferry, rail ferry or density based
+  float factor = (edge->use() == Use::kFerry) * ferry_factor_ +
```

Comment:
Previously the factor was exclusive to either ferry, rail-ferry, or neither.  Now it appears that the logic is written to allow for an edge to be both ferry and rail-ferry - in which case their factors are combined. 

Based on how the code is written tho, I'm guessing edge->use() can only be one value.  (e.g., there aren't any bitmasks involved).

I'm not a huge fan of using the result of the boolean condition as false=0 and true=1 in math below.  How would you feel about using a switch statement?

Reply:
ah yes, i wanted to comment on this, thanks for the reminder! an edge can only have one use, so while i did use addition here only one of the terms can be non-zero, so its really like saying `ferry || rail_ferry || density`

Reply:
i wanted to experiment with what a more drastic removal of branch statements would do to this function as its called very very many times in the main loop of path finding. in my tests the performance wasnt measureably different but i did appreciate the simplification of the code so i left it.

Reply:
Ah!  If performance is the goal then I highly recommend a switch statement.  A switch would work well here.  I think it would also work well in base_transition_cost().

A switch statement compiles to a jump-table, avoiding the cascading conditionals.

Reply:
i dont think we can use it in `base_transition_cost` because value for the switch is likely too complex. we could use it here but im not sure a jump table wouldnt still stall the pipeline anyway. what is suspect is that the branch predictor will realize that most of the time the density based factor will apply and it will make no performance impact. ill give it a quick test, if there is no difference ill just use your suggestion!